[WIP] v6 Docs Dev (jschalz) (#2025)
loft-sh/devspace: docs/pages/configuration/dev/files/sync.mdx

> Make sure that post-sync commands will **<u>not</u>** trigger a new sync process which could lead to an **endless loop**.

Reviewer: ```suggestion
Make sure that post-sync commands will **<u>not</u>** trigger a new sync process . This could lead to an **endless loop**.
```

:::

<<<<<<< HEAD
=======
### `startContainer`

> The `startContainer` option can be used to delay starting a container until sync has finished uploading all files initially. This is very useful if your container entrypoint requires certain files. Needs `command` to be set as otherwise DevSpace doesn't know which command to start.

Reviewer: ```suggestion
The `startContainer` option can be used to delay starting a container until sync has finished uploading all files initially. This is useful if your container entrypoint requires certain files. This option requires `command` to be set; otherwise, DevSpace will not know which command to start.
```

If multiple sync paths with `startContainer` exist, DevSpace will wait for all to finish before starting the container. You can combine `startContainer` sync configurations and non `startContainer` configurations to customize when the container is actually started.
:::

>>>>>>> upstream/v6-docs
### `onUpload.restartContainer`
The `restartContainer` option expects a boolean which defines if DevSpace should restart the container every time either a single file or even a batch of files have been uploaded to the container using file sync.

:::caution Restart Helper Required
<<<<<<< HEAD

> Setting `restartContainer: true` requires to set `injectRestartHelper: true` for the image that is used to run the affected container. Otherwise, this option does not have any effect.

Reviewer: ```suggestion
Setting `restartContainer: true` requires you to also set `injectRestartHelper: true` for the image used to run the affected container. Otherwise, this option has no effect.
```

Author: @LukasGentele, I think I made a mistake in merging and accidentally kept this conflict in. Is this the correct section, or should it be 

```:::caution Restart Helper Required
Setting `restartContainer: true` requires you to also set `command: ["my", "container", "entrypoint"]`. Otherwise, DevSpace doesn't know which command to restart.
:::
```

Reviewer 2: Yes. The quoted section you show here is the correct one. The `injectRestartHelper` option for images is being removed.

=======

> Setting `restartContainer: true` requires to set `command: ["my", "container", "entrypoint"]`. Otherwise, DevSpace doesn't know which command to restart.

Reviewer: ```suggestion
Setting `restartContainer: true` requires you to set `command: ["my", "container", "entrypoint"]`. Otherwise, DevSpace will not know which command to restart.
```

>>>>>>> upstream/v6-docs
:::

:::note When not to use this option

> Using `restartContainer: true` is most useful if your application runs based on a compiled language and you are **not** using a framework or language specific tool which provides hot reloading capabilities. If you are using tools like nodemon or frameworks like React, Rails or Flask, they allow you to enable hot reloading which may be much faster than restarting the entire container. In this case you should disable `restartContainer`.

Reviewer: ```suggestion
Using `restartContainer: true` is most useful if your application runs based on a compiled language and you are **not** using a framework or language-specific tool which provides hot reloading capabilities. If you are using tools like nodemon or frameworks like React, Rails or Flask, they allow you to enable hot reloading, which may be much faster than restarting the entire container. In this case you should disable `restartContainer`.
```

        restartContainer: true
```

>>>>>>> upstream/v6-docs
### `onUpload.exec`

> The `onUpload.exec` option defines command(s) that should be executed after DevSpace uploaded files and folder to the container. DevSpace will make sure that those commands are not executed while initially syncing the state and command execution will also halt any syncing activities.

Reviewer: ```suggestion
The `onUpload.exec` option defines command(s) that should be executed after DevSpace uploaded files and folder to the container. DevSpace will ensure those commands are not executed while initially syncing the state, and command execution will also halt any syncing activities.
```

> If this is defined together with `onUpload.restartContainer`, DevSpace will make sure that the commands are always executed **before** the container is restarted. 

Reviewer: ```suggestion
If this is defined together with `onUpload.restartContainer`, DevSpace will ensure the commands are always executed **before** the container is restarted. 
```

      disableUpload: true
```

>>>>>>> upstream/v6-docs
<br/>

## Initial Sync

<<<<<<< HEAD
=======
### `noWatch`

> If the `noWatch` option is specified, DevSpace will stop syncing after initially comparing and resolving differences. This is useful if you want a one-time sync but then continue doing other steps in your pipeline.

Reviewer: ```suggestion
If the `noWatch` option is specified, DevSpace will stop syncing after initially comparing and resolving differences. This is useful if you want a one-time sync, but then continue doing other steps in your pipeline.
```

<br/>

The `tar` command has to be present in the container otherwise `kubectl cp` does not work and the helper binary cannot be injected into the container.

> Other than that, no server-side component or special container privileges for code synchronization are required, as the sync algorithm runs completely client-only within DevSpace. The synchronization mechanism works with any container filesystem and no special binaries have to be installed into the containers. File watchers running within the containers like nodemon will also recognize changes made by the synchronization mechanism.

Reviewer: ```suggestion
No server-side component or special container privileges for code synchronization are required, since the sync algorithm runs completely client-only within DevSpace. The synchronization mechanism works with any container filesystem and no special binaries have to be installed into the containers. File watchers running within the containers like nodemon will also recognize changes made by the synchronization mechanism.
```

      - node_modules/*
    - path: ./node_modules/:/app/node_modules/
      initialSync: preferRemote
>>>>>>> upstream/v6-docs
```
**Explanation:**  

> With this configuration, `devspace dev` would do the following:

Reviewer: ```suggestion
With this configuration, `devspace dev` performs the following tasks:
```

> - The first sync config section would synchronize all files except files within `node_modules/`. This means that during initial sync, all remote files that are not existing locally would be deleted and other files would be updated to the most recent version.

Reviewer: ```suggestion
- The first sync config section synchronizes all files except files within `node_modules/`. This means that during initial sync, all remote files that do not already exist locally are deleted, and other files are updated to the most recent version.
```

> - The second sync config section would only synchronize files within `node_modules/` and because of `initialSync: preferRemote`, DevSpace would download all remote files which are not present on the local filesystem and override all local files which are different than the files within the container.

Reviewer: ```suggestion
- The second sync config section only synchronizes files within `node_modules/`. Because of `initialSync: preferRemote`, DevSpace downloads all remote files which are not present on the local filesystem and overrides all local files which are different than the files within the container.
```

    sync:
    - path: ./
      waitInitialSync: false
>>>>>>> upstream/v6-docs
```
**Explanation:**  

> With the configuration `devspace dev` would do the following:

Reviewer: ```suggestion
With the configuration `devspace dev` the following actions will be taken:
```

> - DevSpace would start port-forwarding and file synchronzation.

Reviewer: ```suggestion
- DevSpace initiates port-forwarding and file synchronization.
```

> - Initial sync would be started automatically.

Reviewer: ```suggestion
- Initial sync is started automatically.
```

      bandwidthLimits:
        download: 200 # 200 KB/s
        upload: 100   # 100 KB/s
>>>>>>> upstream/v6-docs
```
**Explanation:**  

> - Downloading files from the container to the local filesystem would be limited to a transfer speed of `200 KB/s`.

Reviewer: ```suggestion
- Downloading files from the container to the local filesystem is limited to a transfer speed of `200 KB/s`.
```

## Advanced Options

<<<<<<< HEAD
### `arch`

> Arch specifies which DevSpace helper architecture should be used for the container. Currently valid values are either no value, `amd64` or `arm64`. Depending on this value, DevSpace will inject the DevSpace helper binary with the corresponding architecture suffix.

Reviewer: ```suggestion
Arch specifies which DevSpace helper architecture should be used for the container. Currently valid values are either `no value`, `amd64` or `arm64`. Depending on this value, DevSpace will inject the DevSpace helper binary with the corresponding architecture suffix.
```

=======
>>>>>>> upstream/v6-docs
### `polling`

> Polling specifies if the DevSpace helper should traverse over all watched files and folders periodically in the container to identify file changes. By default, DevSpace will use [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) to detect changes which is more efficient, however sometimes it might be unsupported or not feasible in certain situations, in which polling might be preferred.

Reviewer: ```suggestion
Polling specifies if the DevSpace helper should traverse over all watched files and folders periodically in the container to identify file changes. By default, DevSpace uses [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) to detect changes. This can be more efficient, however, sometimes it might be unsupported or not feasible in certain situations, in which case, polling might be preferred.
```
